File: snippets/design-patterns/structural/adapter.go
// --- Adapter Pattern ---
// Use Case: Allow incompatible interfaces to work together.
// Example: Wrapping a legacy API to conform to a new standard or integrating third-party libraries.
// Description: The Adapter Pattern acts as a bridge between two incompatible interfaces,
// allowing them to work together without modifying their existing code.

package main

import "fmt"

type LegacyPrinter struct{}

func (lp *LegacyPrinter) PrintLegacyMessage() {
	fmt.Println("Printing from LegacyPrinter")
}

type ModernPrinter interface {
	PrintMessage()
}

type Adapter struct {
	legacyPrinter *LegacyPrinter
}

func (a *Adapter) PrintMessage() {
	a.legacyPrinter.PrintLegacyMessage()
}

func main() {
	legacy := &LegacyPrinter{}
	adapter := &Adapter{legacyPrinter: legacy}

	adapter.PrintMessage() // Output: Printing from LegacyPrinter
}

File: snippets/design-patterns/structural/adapter.py
'''
--- Adapter Pattern ---
Use Case: When you need to make two incompatible interfaces work together.
Example: Connecting legacy code to a modern system.
'''

class OldSystem:
    def specific_request(self):
        return "Specific behavior"

class NewSystemAdapter:
    def __init__(self, old_system: OldSystem):
        self.old_system = old_system

    def request(self):
        return self.old_system.specific_request()

# Example Usage
old_system = OldSystem()
adapter = NewSystemAdapter(old_system)
print(adapter.request())  # Output: Specific behavior
